README.md
[![](https://gitlab.com/grammm/php-gram/phpgram/raw/master/docs/img/Feather_writing.svg.png)](https://gitlab.com/grammm/php-gram/phpgram)

# phpgram

[![pipeline status](https://gitlab.com/grammm/php-gram/phpgram/badges/master/pipeline.svg)](https://gitlab.com/grammm/php-gram/phpgram/commits/master)
[![coverage report](https://gitlab.com/grammm/php-gram/phpgram/badges/master/coverage.svg)](https://gitlab.com/grammm/php-gram/phpgram/-/commits/master)
[![Packagist Version](https://img.shields.io/packagist/v/phpgram/phpgram)](https://packagist.org/packages/phpgram/phpgram)
[![PHP from Packagist](https://img.shields.io/packagist/php-v/phpgram/phpgram)](https://gitlab.com/grammm/php-gram/phpgram/blob/master/composer.json)
[![Packagist](https://img.shields.io/packagist/l/phpgram/phpgram)](https://gitlab.com/grammm/php-gram/phpgram/blob/master/LICENSE)

A very fast and lightweight Php Framework for small to enterprise applications.

- RouteHandler for Functions and Classes

- Request and Response via [Psr-7](https://www.php-fig.org/psr/psr-7/) 

- Middleware via [Psr-15](https://www.php-fig.org/psr/psr-15/) 

- [Psr 11](https://www.php-fig.org/psr/psr-11/) Container Support for Automatic Dependency Injection (Autowiring) for Route Callable: Class (in constructor) or Function (with ``__get()``)

- Response Creation (via [Psr-17](https://www.php-fig.org/psr/psr-17/) Response Factory) and Strategies

- **Supports Async Requests**

## Install

Via Composer

``` bash
composer require phpgram/phpgram
```

## [Documentation](https://grammm.gitlab.io/php-gram/phpgram/)

````php
<?php

use Gram\App\App;

use Gram\Strategy\StdAppStrategy;

App::app()->debugMode(true);	//true (default) -> show Exceptions, false -> show empty page

App::app()->setStrategy(new StdAppStrategy()); //define how the Output will be created (StdAppStrategy is default)

App::app()->set404("Er404@404"); //404 Handler
App::app()->set405("Er405@405"); //405 Handler

//Middleware will be executed before the Router
App::app()
	->addMiddleware(new Session())
	->addMiddleware(new Authenticate());


//Routes and Groups

App::app()->get("/","Index@index"); 	//static Route

//dynamic Route with Middleware (this Middleware will only executed if the Routes is matched
App::app()->get("/video/{id:n}","Video@watch")->addMiddleware(new Caching);

//Nested Group with Middleware
// /admin will be the prefix for all Routes in this Group
App::app()->group("/admin",function (){
	
	App::app()->get("","AdminIndexHandler");
	
	App::app()->get("/role","RoleHandler");
	
	//Nested Groups
	//2. Group /settings
	App::app()->group("/settings",function (){
		//3. Group /dep
		App::app()->group("/dep",function (){
			//Routes
			App::app()->get("","DepIndexHandler");
			
			App::app()->get("/{id}/edit","EditHandler");
		});
	})->addMiddleware(new Auth());
	
})->addMiddleware(new Login());

//Group with different Output Strategy (json)
use Gram\Strategy\JsonStrategy;
App::app()->group("/",function (){
	App::app()->getpost("video/vote/{id:n}","Video@vote");
	App::app()->getpost("video/getComment/{id:n}","Video@getComments");
})->addStrategy(new JsonStrategy());


//Psr 7 Request

// in this case with Nyholm Psr 7

use Nyholm\Psr7\Factory\Psr17Factory;
use Nyholm\Psr7Server\ServerRequestCreator;

$psr17Factory = new Psr17Factory();

$creator = new ServerRequestCreator($psr17Factory,$psr17Factory,$psr17Factory,$psr17Factory);
$request = $creator->fromGlobals();

//App needs a Psr 17 Response Factory
App::app()->setFactory($psr17Factory);

//get the response with the content
$response = App::app()->start($request);

//simple emit:
echo $response->getBody(); 	// don't do this in production use an psr 7 emitter!
````

## License

phpgram is open source and under [MIT License](https://gitlab.com/grammm/php-gram/phpgram/blob/master/LICENSE)

## Credits

### Router

- Algorithm and Core Implementation: Copyright by Nikita Popov. ([FastRoute](https://github.com/nikic/FastRoute))

- Parser: Copyright by Nikita Popov and Phil Bennett ([thephpleague](https://github.com/thephpleague/route))

CHANGELOG.md
# CHANGELOG

# 1.8.x

## 1.8.3
- since 2021/03/16
- added php 8 support
- updated dependencies

## 1.8.2

- since 2020/09/23
- DispatcherInterface hat nur noch die function dispatch() (internal bc break)
- Route Object baut nun ihre Middleware von route und von group selber zusammen, das aber nur einmal (internal)
- Route Object ermittelt selbstständig die strategy (internal)
- router gibt nun immer das route object und nicht mehr nur den handler zurück (internal)
- generator trait chunkt nun immer routes, egal was die chunk size zurück gibt. (internal)
- In der App wird die Route Middleware direkt bei build() dem middleware array hinzugefügt (internal)
- queue class hat eine neue function: addMultiple() bei der mehrere mw auf einaml hinzugefügt werden können (internal)
 

## 1.8.1

- since 2020/08/24
- ReadMe Credits wieder hinzugefügt
- GeneratorInterface hat nur die die method generate (internal bc break)
- [#11](https://gitlab.com/grammm/php-gram/phpgram/-/issues/11) Middleware und Strategy werden mit einem static Array gesammelt und nicht mehr mit einem extra Collector

## 1.8.0

- since 2020/08/07
- Std middleware werden nun direkt in der App class gesammelt und nicht mehr im Collector
- [#7](https://gitlab.com/grammm/php-gram/phpgram/-/issues/7) Bei einem aufruf von Route group wird der collector mitgegeben
- [#9](https://gitlab.com/grammm/php-gram/phpgram/-/issues/9) Alle Attribute im Request werden nun durch const's repräsentiert
- [#6](https://gitlab.com/grammm/php-gram/phpgram/-/issues/6) Der QueueHandler lässt sich nun auch als callable aufrufen
- [#8](https://gitlab.com/grammm/php-gram/phpgram/-/issues/8) [bc break!] Emitter wird nicht mehr unterstüzt. Bei `start()` wird nur noch ein fertiges ResponseInterface zurück gegeben

# 1.7.x

## 1.7.0

- since 2020/02/20
- Router Bug fix:
	- Unterstüzt jetzt auch optionale Parameter
	- nun kann bei den Static Routes (ohne Parameter) auch Urls mit einem . gematcht werden
	- Route Data wird dem Dispatcher im Constructor übergeben
- App Constructor ist nun nicht mehr private
- QueueHandler ist nun auch ein Interface

# 1.6.x

## 1.6.2
- since 2020/01/27
- Memory Leaks beim QueueHandler fix

## 1.6.1
- since 2020/01/26
- Dispatcher Warning wenn keine Routes gesetzt sind entfernt

## 1.6.0

- since 2020/01/23
- Middleware werden nun in einem Queue Object gespeichert	
	- dieses wird bei jedem Request erstellt
- Async App entfernt, da die normale App somit auch Async Ready ist
- rawOptions in der App Class entfernt
- Debug Mode nur noch auf true oder false setzen:
	- bei true -> render exception
	- bei false -> nur Response mit 500 zurück geben
- Bei 404 bzw. 405 wird nun der richtige Statuscode bei Exceptions ausgegeben, sollte kein Handler gesetzt sein
- Resolver müssen nun bei resolve() ein array mit den Paramtern bekommen


# 1.5.x

## 1.5.3

- since 2020/01/18
- Routing Middleware:
	- Bereiche mit Methods aufgeteilt, damit die Async Mw nicht alles doppelt machen muss
- Dispatcher: 
	- bei 405 bzw. 404 werden doppelte Http Methods nur noch einmal durchsucht
- QueueHandler unterstüzt nun auch callable Middleware
- Interfaces angepasst
- adMiddle() zu addMiddleware() umbenannt

## 1.5.2
- since 2020/01/13
- CallableResolver aufgeteilt in ClosureResolver und CallableResolver
	- ClosureResolver ist für anonymous functions
	- CallableResolver für alle anderen callable, wie Klassen mit der Method __invoke()
- App gibt jetzt auch den Psr 11 Container wieder zurück
- Router: als Standard Generator und Dispatcher werden nun MarkBased verwendet

## 1.5.1
- since 2020/01/03
- Router:
	- Es wird jetzt nur noch MethodSort genutzt
	- Es wird in der Routemap nur noch die Route Id gespeichert. (dadurch ist diese kleiner und verbraucht weniger Ram)
	- Router holt sich die Infos aus dem Routeobject
	- MarkBased generator und dispatcher hinzugefügt

## 1.5.0
- since 2019/12/30
- Async Requests:
	 - Es werden jetzt auch ohne eine Erweitertung Async Requests unterstüzt
	 - Dazu einfach die AsyncApp anstatt der normalen App Class verwenden
- Route:
	- Parsen der Route findet nun im Generator und nicht mehr direkt in der Route statt
	- Es wurde ein allgemeiner Generator dafür erstellt
	- Die normalen Generatoren und Dispatcher sind nun im Ordner Std
- QueueHandler ist nun im Middleware Ordner und im Middleware Namespace
- Allgemein Code Quality verbessert
- Route Collector: addGroup() wurde zu group() umbenannt
- Strategy:
	- Erstellt nun den Response
	- Ein Response mit string Content kann mithilfe eines Traits erstellt werden
	- Für alle anderen Inhalte muss entweder eine neue Strategy geschrieben werden oder der Response direkt zurückgegeben werden
	- Buffer Strategy kann nun auch einen Response aus dem Buffer zurück geben

# 1.4.x

## 1.4.2
- since 2019/12/14
- Möglichkeit Exceptions an zuzeigen eingebaut. Nun wird nur noch wenn gewünscht die Exception ausgegeben
- Strategies können nun auch aus dem Container geladen werden

## 1.4.1
- since 2019/12/09
- Es kann nun auch auf ResponseCreator, zur Laufzeit, zugegriffen werden, mit `getResponseCreator()`
	- Vorher müssen aber die Factories gesetzt sein
	- Sollte eine andere Strategy verwendet werden muss diese auch vorher gesetzt sein
	- Lösung: ResponseCreator in den Psr 11 Container legen und alle Middleware die diesen brauchen ebenfalls in den Container
	- ResponseCreator wird zur Laufzeit erstellt in der App, dieser Container sollte das ebenfalls mit dem ResponseCreator tun
	
## 1.4.0
- since 2019/12/06
- neuer Route Generator und Dispatcher hinzugefügt
	- kann optional genutzt werden
	- Generator ordnet Routes bei ihrer Http Method ein. 
	- Sollte eine Route mehrere Http Methods haben wird diese mehrfach in der Routemap mit ihrer Id und Group Id auftauchen
	- gut geeignet für opcache
- NotFoundHandler wirft nun Exceptions, wenn kein Handler angegeben wurde


# 1.3.x

## 1.3.1
- since 2019/11/05
- ResolverCreator kann nur noch am Anfang gesetzt werden
- Middleware Order (außer Betrieb) entfernt


src/Route/Router.php
<?php
/**
 * phpgram
 *
 * This File is part of the phpgram Micro Framework
 *
 * Web: https://gitlab.com/grammm/php-gram/phpgram
 *
 * @license https://gitlab.com/grammm/php-gram/phpgram/blob/master/LICENSE
 *
 * @author Jörn Heinemann <joernheinemann@gmx.de>
 */

namespace Gram\Route;

use Gram\Route\Interfaces\CollectorInterface;
use Gram\Route\Interfaces\DispatcherInterface;
use Gram\Route\Interfaces\RouterInterface;

/**
 * Class Router
 * @package Gram\Route
 *
 * Der Router der von der Routingmiddleware aufgerufen wird
 *
 * Kann auch ohne Psr 7 genutzt werden
 *
 * Verschiedene Optionen für Dispatcher, Generator, Parser und Collector sind setztbar
 *
 * Führt das dispatching aus
 */
class Router implements RouterInterface
{
	/** @var bool */
	private $slash_trim;

	/** @var CollectorInterface */
	private $collector;

	/** @var DispatcherInterface */
	private $dispatcher;

	/** @var bool */
	protected $caching;

	/** @var null|string */
	protected $cache;

	/**
	 * Router constructor.
	 * @param array $options
	 */
	public function __construct($options = [])
	{
		//setze Standard Optionen
		$options +=[
			'slash_trim'=>true,
			'caching'=>false,
			'cache'=>null,
			'dispatcher'=>'Gram\\Route\\Dispatcher\\MarkBased',
			'generator'=>'Gram\\Route\\Generator\\MarkBased',
			'parser'=>'Gram\\Route\\Parser\\StdParser',
			'collector'=>'Gram\\Route\\Collector\\RouteCollector'
		];

		$this->slash_trim = $options['slash_trim'];

		$this->caching = $options['caching'];

		$this->cache = $options['cache'];

		//Erstelle den Collector, der wird auch für andere Klassen verfügbar sein
		$this->collector= new $options['collector'](
			new $options['generator'](new $options['parser'])
		);

		$this->dispatcher = $options['dispatcher'];	//Dispatcher
	}

	/**
	 * @inheritdoc
	 */
	public function run($uri,$httpMethod)
	{
		$uri = \urldecode($uri);	//umlaute filtern

		if($this->slash_trim && $uri !== $this->collector->getBase().'/'){
			$uri = \rtrim($uri,'/');	//entferne letzen / von der Url
		}

		if(\is_string($this->dispatcher)) {
			//sollte der Dispatcher noch nicht erstellt sein
			$this->dispatcher = new $this->dispatcher($this->getData());
		}

		$response = $this->dispatcher->dispatch($httpMethod,$uri);

		return $this->getHandle($response);
	}

	/**
	 * Prüfe zuerst ob es einen Cache für die generierten Routes gibt
	 *
	 * Wenn ja lade den Cache
	 *
	 * Sonst hole die generierten Routes von dem RouteCollector
	 *
	 * @return array
	 */
	protected function getData()
	{
		if($this->caching && \file_exists($this->cache)){
			return require $this->cache;
		}

		$data = $this->collector->getData();

		if($this->caching) {
			\file_put_contents(
				$this->cache,
				'<?php return ' . \var_export($data, true) . ';'
			);
		}

		return $data;
	}

	/**
	 * Gibt den Handle der Route zurück
	 *
	 * bei 404 oder 405 den jeweiligen Handle
	 *
	 * @param array $response
	 * @return array
	 */
	protected function getHandle(array $response)
	{
		if($response[0] === DispatcherInterface::FOUND) {
			$route = $this->collector->getRoute($response[1]);

			return [$response[0],$route,$response[2]];
		}

		if($response[0] === DispatcherInterface::NOT_ALLOWED){
			 $handle = $this->collector->get405();
		}else {
			$handle = $this->collector->get404();
		}

		return [$response[0],$handle,[]];
	}

	/**
	 * @inheritdoc
	 */
	public function getCollector()
	{
		return $this->collector;
	}
}